global_helpers/panther_base_helpers.py

```diff
@@ -494,3 +495,25 @@ def m365_alert_context(event):
 def defang_ioc(ioc):
     """return defanged IOC from 1.1.1.1 to 1[.]1[.]1[.]1"""
     return ioc.replace(".", "[.]")
+
+def unmock(func: callable, *args, **kwargs):
+    """Wraps mocked functions to perform automatic JSON unmarshalling on the returned value during
+    unit tests, and return the unmodified value during actual log analysis.
+    
+    Usage:
+        func: the mocked function you want to call
+        *args: the positional arguments of 'func', if any
+        **kwargs: the keyword arguments of 'func', if any
+    
+    Example:
+        To wrap `get_counter("cache-key", force_ttl=True)`, call:
+            unmock(get_counter, "cache-key", force_ttl=True)
+    """
+    # Call the function and save the return value
+    val = func(*args, **kwargs)
+
+    # If this is a mocked function, unmarhsal the returned string as JSON
+    if isinstance(func, MagicMock):
```

Comment:
I don't love turning this implementation detail (the type of our mocks) into a de-facto interface by exposing it here. Looks like the example in the [docs](https://docs.panther.com/detections/testing#example-mock) assumes that the returned value shouldn't be a string, which has more limited uses but doesn't expose the mock details.

Hmm, is this maybe something we want to write as part of panther-detection-helpers instead, since that has less public exposure?

Reply:
We have a few rules already that do this, which is where I got the idea. If we don't want to reveal this class, we should probably edit those rules too. Can you explain what panther-detection-helpers is? I'm not familiar with it

Reply:
> If we don't want to reveal this class, we should probably edit those rules too.

Not revealing it is definitely best practice, but if we're already revealing it, I don't have as strong opinions about whether we do so here

> Can you explain what panther-detection-helpers is?

It's a new private repository used by the detections engine - it's where we moved the kv-store code to, so that we can edit and iterate on it without customers needing to update panther-analysis to get each change

Reply:
Ah, thanks for that info! I didn't know that's where we moved that to. Re: what to do about the Magic Mock, is the concern that we might move away from the MagicMock data type one day, or is it moreso "magicians don't reveal their secrets"?

Reply:
It's more the former - we've unnecessarily locked ourselves into that implementation and/or made it much harder to move away from it, should we ever want to

Reply:
Based on a quick glance at `panther_detection_helpers`, I get the impression that it isn't something we'd expect customers to import from directly, so I'm thinking the way to do this would be to implement 2 functions:

In `panther_detection_helpers`, we can add `is_mock`, that checks if a callable is an instance of `MagicMock`. We can change the implementation without it being visible to customers. Here, in `panther_base_helpers`, we can use this `unmock` function, but import `is_mock` and use that for the if-statement.

Would that work for you?

Reply:
I'd be fine with either that approach or the current one

Reply:
I think the splitting approach makes more sense. Is there any documentation on how `panther_detection_helpers` should be structured? Do we want everything to be in a sub-module, or should I add `is_mock` to the root?

Reply:
No docs, it's pretty new. @maxrichie5 do you have thoughts on structure here?

Reply:
Definitely add it in a submodule. Possibly `mocking`

Also not blocking to this PR: but this really should be a feature improvement of how we support mocking instead of a global. So this should ideally be a temp solution

Reply:
Sounds good. I agree, this is a temp measure until we expand mocking support to other data types! But, since this is a comparatively quick and easy fix, figured it's better to have something in place.